File: web/src/main/kotlin/cloud/fabX/fabXaccess/tool/rest/Tool.kt
package cloud.fabX.fabXaccess.tool.rest

import cloud.fabX.fabXaccess.common.rest.ChangeableValue
import kotlinx.serialization.Serializable

@Serializable
data class Tool(
    val id: String,
    val aggregateVersion: Long,
    val name: String,
    val type: ToolType,
    val requires2FA: Boolean,
    val time: Int,
    val idleState: IdleState,
    val enabled: Boolean,
    val wikiLink: String,
    val requiredQualifications: Set<String>
)

fun cloud.fabX.fabXaccess.tool.model.Tool.toRestModel(): Tool = Tool(
    id = id.serialize(),
    aggregateVersion = aggregateVersion,
    name = name,
    type = type.toRestModel(),
    requires2FA = requires2FA,
    time = time,
    idleState = idleState.toRestModel(),
    enabled = enabled,
    wikiLink = wikiLink,
    requiredQualifications = requiredQualifications.map { it.serialize() }.toSet()
)

enum class ToolType {
    UNLOCK,
    KEEP
}

fun cloud.fabX.fabXaccess.tool.model.ToolType.toRestModel(): ToolType {
    return when (this) {
        cloud.fabX.fabXaccess.tool.model.ToolType.UNLOCK -> ToolType.UNLOCK
        cloud.fabX.fabXaccess.tool.model.ToolType.KEEP -> ToolType.KEEP
    }
}

fun ToolType.toDomainModel(): cloud.fabX.fabXaccess.tool.model.ToolType {
    return when (this) {
        ToolType.UNLOCK -> cloud.fabX.fabXaccess.tool.model.ToolType.UNLOCK
        ToolType.KEEP -> cloud.fabX.fabXaccess.tool.model.ToolType.KEEP
    }
}

enum class IdleState {
    IDLE_LOW,
    IDLE_HIGH
}

fun cloud.fabX.fabXaccess.tool.model.IdleState.toRestModel(): IdleState {
    return when (this) {
        cloud.fabX.fabXaccess.tool.model.IdleState.IDLE_LOW -> IdleState.IDLE_LOW
        cloud.fabX.fabXaccess.tool.model.IdleState.IDLE_HIGH -> IdleState.IDLE_HIGH
    }
}

fun IdleState.toDomainModel(): cloud.fabX.fabXaccess.tool.model.IdleState {
    return when (this) {
        IdleState.IDLE_LOW -> cloud.fabX.fabXaccess.tool.model.IdleState.IDLE_LOW
        IdleState.IDLE_HIGH -> cloud.fabX.fabXaccess.tool.model.IdleState.IDLE_HIGH
    }
}

@Serializable
data class ToolCreationDetails(
    val name: String,
    val type: ToolType,
    val requires2FA: Boolean,
    val time: Int,
    val idleState: IdleState,
    val wikiLink: String,
    val requiredQualifications: Set<String>
)

@Serializable
data class ToolDetails(
    val name: ChangeableValue<String>?,
    val type: ChangeableValue<ToolType>?,
    val requires2FA: ChangeableValue<Boolean>?,
    val time: ChangeableValue<Int>?,
    val idleState: ChangeableValue<IdleState>?,
    val enabled: ChangeableValue<Boolean>?,
    val wikiLink: ChangeableValue<String>?,
    val requiredQualifications: ChangeableValue<Set<String>>?
)

File: frontend-svelte/src/lib/api/model/tool.ts
import type { ChangeableValue } from './changeable-value';
import type { Qualification } from '$lib/api/model/qualification';

export interface Tool {
	id: string;
	aggregateVersion: number;
	name: string;
	type: ToolType;
	requires2FA: boolean;
	time: number;
	idleState: IdleState;
	enabled: boolean;
	notes: string | null;
	wikiLink: string;
	requiredQualifications: string[];
}

export interface AugmentedTool {
	id: string;
	aggregateVersion: number;
	name: string;
	type: ToolType;
	requires2FA: boolean;
	time: number;
	idleState: IdleState;
	enabled: boolean;
	notes: string | null;
	wikiLink: string;
	requiredQualifications: Qualification[];
}

export interface ToolCreationDetails {
	name: string;
	type: ToolType;
	requires2FA: boolean;
	time: number;
	idleState: IdleState;
	wikiLink: string;
	requiredQualifications: string[];
}

export interface ToolDetails {
	name: ChangeableValue<string> | null;
	type: ChangeableValue<ToolType> | null;
	requires2FA: ChangeableValue<boolean> | null;
	time: ChangeableValue<number> | null;
	idleState: ChangeableValue<IdleState> | null;
	enabled: ChangeableValue<boolean> | null;
	notes: ChangeableValue<string | null> | null;
	wikiLink: ChangeableValue<string> | null;
	requiredQualifications: ChangeableValue<string[]> | null;
}

export const toolTypes = ['UNLOCK', 'KEEP'];
export type ToolType = (typeof toolTypes)[number];

export const idleStates = ['IDLE_LOW', 'IDLE_HIGH'];
export type IdleState = (typeof idleStates)[number];
